### PR TITLE
Add zone volume creation integration tests

### PR DESCRIPTION
**Pull request by falfaroc**
# Description
Add and separate the zone volume integration test. Ensured that the initial integration tests are still able to run with the original commands.

Additionally, create new `e2e` specific tests that can be run to test the driver while already configured and running. Since the zoning functionality depends on the CO of k8s to make the appropriate decision, verification with a live cluster is ideal.

**Note:** Reason for the separation is because of the array configuration needed at the start and I want to ensure that we can still run the original integration tests.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1612 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Run and ensure that the zone integration tests are passing.
```
...
=== RUN   TestZoneIntegration/Create_publish,_unpublish,_and_delete_zone_volume

  Scenario: Create publish, unpublish, and delete zone volume # features/integration.feature:985
    Given a VxFlexOS service                                  # <autogenerated>:1 -> *feature
    And I create a zone volume request "zone-integration-vol" # <autogenerated>:1 -> *feature
INFO[0005] Preferred topology zone zoneA, systemID 205f819048a7000f, protectionDomain domain1, and storagePool pool1
INFO[0005] [CreateVolume] Multi-AZ Storage Pool Determined by Secret pool1
INFO[0005] Executing CreateVolume with following fields  name=zone-integration-vol_101321 sizeInKiB=8388608 storagePool=pool1 volType=ThinProvisioned x-csi-pv-claimname= x-csi-pv-name= x-csi-pv-namespace=
INFO[0005] volume zone-integration-vol_101321 (205f819048a7000f-01f6e8f80000002c) created 2024-11-29 10:10:22 -0500 EST
CreateVolume zone-integration-vol_101321 (205f819048a7000f-01f6e8f80000002c) 2024-11-29 10:10:22 -0500 EST
    When I call CreateVolume                                  # <autogenerated>:1 -> *feature
    And there are no errors                                   # <autogenerated>:1 -> *feature
req.NodeId 323E2A19-FC01-46D4-AFA9-3CC9AD92811E
INFO[0006] Volume ID: 205f819048a7000f-01f6e8f80000002c contains system ID: 205f819048a7000f. checkVolumesMap passed
ControllerPublishVolume completed successfully
    And when I call PublishVolume "SDC_GUID"                  # <autogenerated>:1 -> *feature
    And there are no errors                                   # <autogenerated>:1 -> *feature
[FERNANDO] Waiting 20 seconds before unpublishing...
[FERNANDO] Let's unpublish...
INFO[0026] Volume ID: 205f819048a7000f-01f6e8f80000002c contains system ID: 205f819048a7000f. checkVolumesMap passed
ControllerUnpublishVolume completed successfully
    And when I call UnpublishVolume "SDC_GUID"                # <autogenerated>:1 -> *feature
    And there are no errors                                   # <autogenerated>:1 -> *feature
INFO[0027] Volume ID: 205f819048a7000f-01f6e8f80000002c contains system ID: 205f819048a7000f. checkVolumesMap passed
INFO[0027] Deleting volume                               id=205f819048a7000f-01f6e8f80000002c name=zone-integration-vol_101321
DeleteVolume 205f819048a7000f-01f6e8f80000002c completed successfully
    And when I call DeleteVolume                              # <autogenerated>:1 -> *feature
    Then there are no errors                                  # <autogenerated>:1 -> *feature
=== RUN   TestZoneIntegration/Create_zone_volume_with_invalid_zone_information

  Scenario: Create zone volume with invalid zone information # features/integration.feature:998
INFO[0027] no zone target for {  }
CreateVolume rpc error: code = InvalidArgument desc = no zone topology found in accessibility requirements:
    Given a VxFlexOS service                                 # <autogenerated>:1 -> *feature
    And I create an invalid zone volume request              # <autogenerated>:1 -> *feature
    When I call CreateVolume                                 # <autogenerated>:1 -> *feature
    Then the error message should contain <errormsg>         # <autogenerated>:1 -> *feature

    Examples:
      | errormsg                                               |
      | "no zone topology found in accessibility requirements" |

2 scenarios (2 passed)
14 steps (14 passed)
22.269408149s
INFO[0027] gracefully stopped
--- PASS: TestZoneIntegration (27.29s)
    --- PASS: TestZoneIntegration/Create_publish,_unpublish,_and_delete_zone_volume (22.27s)
    --- PASS: TestZoneIntegration/Create_zone_volume_with_invalid_zone_information (0.00s)
PASS
coverage: [no statements]
ok      github.com/dell/csi-vxflexos/v2/test/integration        27.336s
```
- [x] Run e2e tests for zone scheduling
```
$ make zone-e2e
go test -v -count=1 -timeout 1h -run '^TestZoneVolumes$' ./e2e
=== RUN   TestZoneVolumes
Feature: VxFlex OS CSI interface
  As a consumer of the CSI interface
  I want to run a system test
  So that I know the service functions correctly.
=== RUN   TestZoneVolumes/Create_zone_volume_through_k8s

  Scenario: Create zone volume through k8s                                    # features/e2e.feature:7
2024/12/03 16:55:05 [isEverythingWorking] Checking if everything is working...
2024/12/03 16:55:19 [createZoneVolumes] Created volumes and pods...
2024/12/03 16:55:19 [checkStatfulSetStatus] checking statefulset status
2024/12/03 16:56:42 [deleteZoneVolumes] Deleted volumes and pods...
    Given a VxFlexOS service                                                  # <autogenerated>:1 -> *feature
    And verify driver is configured and running correctly                     # <autogenerated>:1 -> *feature
    And verify zone information from secret <secret> in namespace <namespace> # <autogenerated>:1 -> *feature
    Then create zone volume and pod in <location>                             # <autogenerated>:1 -> *feature
    And check the statefulset for zones                                       # <autogenerated>:1 -> *feature
    Then delete zone volume and pod in <location>                             # <autogenerated>:1 -> *feature

    Examples:
      | secret            | namespace  | location    |
      | "vxflexos-config" | "vxflexos" | "zone-wait" |
=== RUN   TestZoneVolumes/Cordon_node_and_create_zone_volume_through_k8s

  Scenario: Cordon node and create zone volume through k8s                    # features/e2e.feature:19
2024/12/03 16:56:42 [isEverythingWorking] Checking if everything is working...
2024/12/03 16:56:45 [cordonNode] Cordoning node: worker-1-njo3bz3jjysjb
2024/12/03 16:56:46 [cordonNode] Cordoned node correctly
2024/12/03 16:56:57 [createZoneVolumes] Created volumes and pods...
2024/12/03 16:56:57 [checkStatfulSetStatus] checking statefulset status
2024/12/03 16:58:21 [checkPodsForCordonRun] checking pods status
2024/12/03 16:58:22 [getStatefulSetPods] Pods found:  7
2024/12/03 16:58:22 [checkPodsForCordonRun] Pods scheduled correctly, reseting node...
2024/12/03 16:58:33 [deleteZoneVolumes] Deleted volumes and pods...
    Given a VxFlexOS service                                                  # <autogenerated>:1 -> *feature
    And verify driver is configured and running correctly                     # <autogenerated>:1 -> *feature
    And verify zone information from secret <secret> in namespace <namespace> # <autogenerated>:1 -> *feature
    Then cordon one node                                                      # <autogenerated>:1 -> *feature
    Then create zone volume and pod in <location>                             # <autogenerated>:1 -> *feature
    And check the statefulset for zones                                       # <autogenerated>:1 -> *feature
    And ensure pods aren't scheduled incorrectly and still running            # <autogenerated>:1 -> *feature
    Then delete zone volume and pod in <location>                             # <autogenerated>:1 -> *feature

    Examples:
      | secret            | namespace  | location    |
      | "vxflexos-config" | "vxflexos" | "zone-wait" |

2 scenarios (2 passed)
14 steps (14 passed)
3m28.084932097s
--- PASS: TestZoneVolumes (208.09s)
    --- PASS: TestZoneVolumes/Create_zone_volume_through_k8s (97.72s)
    --- PASS: TestZoneVolumes/Cordon_node_and_create_zone_volume_through_k8s (110.36s)
PASS
ok      github.com/dell/csi-vxflexos/v2/test/e2e        208.119s
```

### CODE REVIEW

File: test/e2e/features/e2e.feature [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    Given a VxFlexOS service
+    And verify driver is configured and running correctly
+    And verify that the node and storage class zone label is <zoneLabelKey>
+    And verify that the secret <secret> in namespace <namespace> is set for zoning
+    Then create zone volume and pod in <location>
+    And check pods to be running on desired zones
+    Then delete zone volume and pod in <location>
+    Examples:
+      | zoneLabelKey                    | secret            | namespace  | location    |
+      | "zone.csi-vxflexos.dellemc.com" | "vxflexos-config" | "vxflexos" | "zone-wait" |
```
Comment (tdawe) [inline test/e2e/features/e2e.feature:17]: Can we add an additional case here for a custom zone label?

Comment (falfaroc) [reply test/e2e/features/e2e.feature:17]: Yup, I was waiting for the other PR to add anything related to zoneLabelKey.

Comment (falfaroc) [reply test/e2e/features/e2e.feature]: Removed `zoneLabelKey` argument. E2E assumes that the driver is running and setup correctly so we are now retrieving the zone label from the secret and ensure that the label key is consistent across the nodes and storage class.

**Review by tdawe — APPROVED** [review]

File: test/integration/run.sh [hunk 2]
```diff
@@ -23,7 +23,8 @@ if [ $rc -ne 0 ]; then echo "failed http unauthorized test"; exit $rc; fi
 rm -f unix_sock
 . ../../env.sh
 echo $SDC_GUID
-GOOS=linux CGO_ENABLED=0 GO111MODULE=on go test -v -coverprofile=c.linux.out -timeout 60m -coverpkg=github.com/dell/csi-vxflexos/service *test.go &
+
+GOOS=linux CGO_ENABLED=0 GO111MODULE=on go test -v -coverprofile=c.linux.out -timeout 60m -coverpkg=github.com/dell/csi-vxflexos/service -run "^$1\$\$" &
```
Comment (lukeatdell) [inline test/integration/run.sh:27]: can we name the `$1` var?

File: test/integration/features/array-config/multi-az [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+   {
+      "endpoint": "https://127.0.0.1",
+      "username": "username",
+      "password": "password",
+      "insecure": true,
+      "isDefault": true,
+      "systemID": "0000000000000001",
+      "mdm": "127.0.0.2,127.0.0.3",
+      "zone": {
+         "name": "myZone",
```
Comment (lukeatdell) [inline test/integration/features/array-config/multi-az:11]: does this need to be updated to add `labelKey`?

**Review by xuluna — APPROVED** [review]

**Review by lukeatdell — APPROVED** [review]
just a couple small comments, otherwise LGTM

File: test/e2e/templates/zone-wait/pod.yaml [hunk 4]
```diff
@@ -0,0 +1,200 @@
+---
```
Comment (bharathsreekanth) [inline test/e2e/templates/zone-wait/pod.yaml:1]: Do we need to create individual PVC and pods to consume them, or instead use a statefulset like what Luke achieved and dynamically create them and consume them?

File: test/integration/run.sh (hunk 2 above)
Comment (falfaroc) [reply test/integration/run.sh:27]: I can do that 🫡

File: test/e2e/templates/zone-wait/pod.yaml (hunk 4 above)
Comment (falfaroc) [reply test/e2e/templates/zone-wait/pod.yaml:1]: I can convert it to the statefulset, I just have to rework the checks to ensure that everything is up and running for the pods. Will take a look.

**Review by lukeatdell — APPROVED** [review]

**Pull request merged**